### ISSUE TITLE
Block Handling

### ISSUE THREAD

**Issue by stanhebben** [NOVAAPI/NovaSuggestions#9]
I notice that blocks currently contain their position and blockAccess. From a design perspective, this is good, since you wouldn't have to pass these parameters for every function you want to call on it.

However, this means that blocks cannot be reused, even if they don't have any state. Since in a typical world, most common blocks don't have state (such as stone, dirt, grass, ...). Each of them would either need to be instantiated when retrieved from the BlockAccess, or they would have to be stored somehow.

What if blocks didn't store their position, much in the same way Minecraft doesn't, unless they explicitly choose to? Neither would blocks receive their position as parameters for their functions - if location matters, they should store position and blockAccess themselves.

Blocks could register themselves in two ways. Blocks without state and without need for world access (which is, any simple decorative block without connecting textures) could just register a Block instance under a given name (their ID). Whenever a block is needed of that ID, it can simply return the Block instance.

Alternatively, if position, state and/or world access does matter for the block, a BlockFactory can be registered. This factory is called (with world access, position and block data) when the block needs to be instantiated. If desired, the factory can still choose to return the same block multiple times if they are somehow equivalent. (this could allow for large, efficient microblock structures if block contents are the same for many blocks)

**Comment by calclavia** [comment]
Would you say, based on yesterday's changes, that this is resolved?

**Comment by stanhebben** [comment]
It was before you changed it :D

I don't feel confident with the state it is in now, since there is only one mechanism. It may not cover all ways modders want to use it. I believe the construct & reflect the data in should be an option, but not the only one. (Currently reusing the same block instance is impossible, again, thus being a step backward imo)

The system that was present before these changes allowed for more flexibility in the way blocks are constructed (and possibly reused), and also allowed us to use the reflection - based system. Now it always uses the reflected system. I believe an API should offer flexibility.

When we start to provide block data for block deserialization (for stateful blocks), I am worried that the system will work against us. Extending the PositionedConstructor interface with an additional block data parameter would have made it easy to provide block data to it, and implementations could even choose a different block class depending on the data. This is not possible now.

Nonetheless, the reflection-based initialization needs little code from the modder, and I am convinced that the option is very useful, just that it shouldn't be the only way.

**Comment by calclavia** [comment]
Yes, that's why we still have that lambda constructor. That'll be for custom constructors. Is that not enough?

**Comment by Kubuxu** [comment]
@stanhebben Block cannot be reused as we plan to introduce WeakValueMap<Position,Block>  to save on creation overhead

**Comment by stanhebben** [comment]
Corrected my comment. Block reuse is indeed possible using the lambda constructor.

Using a cache doesn't exclude the option of reusing stateless blocks - a cache can contain the same object values for different positions.
